In [31]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from pydataset import data

Attendance Data

Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [32]:
attendance = pd.read_csv('untidy-data/attendance.csv')

In [33]:
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [34]:
attendance = attendance.rename(columns={'Unnamed: 0': 'student_name'})

In [35]:
attendance

student_name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0        Sally          P          T          T          H          P   
1         Jane          A          P          T          T          T   
2        Billy          A          T          A          A          H   
3         John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [36]:
attendance = attendance.melt(id_vars='student_name')


In [37]:
attendance

student_name    variable value
0         Sally  2018-01-01     P
1          Jane  2018-01-01     A
2         Billy  2018-01-01     A
3          John  2018-01-01     P
4         Sally  2018-01-02     T
5          Jane  2018-01-02     P
6         Billy  2018-01-02     T
7          John  2018-01-02     T
8         Sally  2018-01-03     T
9          Jane  2018-01-03     T
10        Billy  2018-01-03     A
11         John  2018-01-03     H
12        Sally  2018-01-04     H
13         Jane  2018-01-04     T
14        Billy  2018-01-04     A
15         John  2018-01-04     P
16        Sally  2018-01-05     P
17         Jane  2018-01-05     T
18        Billy  2018-01-05     H
19         John  2018-01-05     P
20        Sally  2018-01-06     A
21         Jane  2018-01-06     T
22        Billy  2018-01-06     T
23         John  2018-01-06     T
24        Sally  2018-01-07     T
25         Jane  2018-01-07     A
26        Billy  2018-01-07     P
27         John  2018-01-07     P
28        Sally  2018-01-08     T
29         Jane  2018-01-08     T
30        Billy  2018-01-08     T
31         John  2018-01-08     P

In [38]:
 attendance.drop(columns='variable')

student_name value
0         Sally     P
1          Jane     A
2         Billy     A
3          John     P
4         Sally     T
5          Jane     P
6         Billy     T
7          John     T
8         Sally     T
9          Jane     T
10        Billy     A
11         John     H
12        Sally     H
13         Jane     T
14        Billy     A
15         John     P
16        Sally     P
17         Jane     T
18        Billy     H
19         John     P
20        Sally     A
21         Jane     T
22        Billy     T
23         John     T
24        Sally     T
25         Jane     A
26        Billy     P
27         John     P
28        Sally     T
29         Jane     T
30        Billy     T
31         John     P

In [39]:
attendance = attendance.drop(columns='variable')

In [40]:
attendance

student_name value
0         Sally     P
1          Jane     A
2         Billy     A
3          John     P
4         Sally     T
5          Jane     P
6         Billy     T
7          John     T
8         Sally     T
9          Jane     T
10        Billy     A
11         John     H
12        Sally     H
13         Jane     T
14        Billy     A
15         John     P
16        Sally     P
17         Jane     T
18        Billy     H
19         John     P
20        Sally     A
21         Jane     T
22        Billy     T
23         John     T
24        Sally     T
25         Jane     A
26        Billy     P
27         John     P
28        Sally     T
29         Jane     T
30        Billy     T
31         John     P

In [41]:
attendance['value'] = attendance['value'].replace(['P','A','T','H'], [1,0,.9,.5])

In [42]:
attendance

student_name  value
0         Sally    1.0
1          Jane    0.0
2         Billy    0.0
3          John    1.0
4         Sally    0.9
5          Jane    1.0
6         Billy    0.9
7          John    0.9
8         Sally    0.9
9          Jane    0.9
10        Billy    0.0
11         John    0.5
12        Sally    0.5
13         Jane    0.9
14        Billy    0.0
15         John    1.0
16        Sally    1.0
17         Jane    0.9
18        Billy    0.5
19         John    1.0
20        Sally    0.0
21         Jane    0.9
22        Billy    0.9
23         John    0.9
24        Sally    0.9
25         Jane    0.0
26        Billy    1.0
27         John    1.0
28        Sally    0.9
29         Jane    0.9
30        Billy    0.9
31         John    1.0

In [43]:
 attendance.groupby('student_name').mean()

value
student_name        
Billy         0.5250
Jane          0.6875
John          0.9125
Sally         0.7625

In [44]:
attendance = attendance.groupby('student_name').mean()

In [45]:
attendance

value
student_name        
Billy         0.5250
Jane          0.6875
John          0.9125
Sally         0.7625

Coffee Levels

Read the coffee_levels.csv file.
Transform the data so that each carafe is in it's own column.
Is this the best shape for the data?

In [46]:
coffee_levels = pd.read_csv('untidy-data/coffee_levels.csv')

In [47]:
coffee_levels.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [48]:
coffee_levels.coffee_carafe.value_counts()

x    10
y    10
z    10
Name: coffee_carafe, dtype: int64

In [49]:
new_coffee_levels = coffee_levels.pivot(index='hour', columns='coffee_carafe')

In [50]:
new_coffee_levels

coffee_amount                    
coffee_carafe             x         y         z
hour                                           
8                  0.816164  0.189297  0.999264
9                  0.451018  0.521502  0.915990
10                 0.843279  0.023163  0.144928
11                 0.335533  0.235529  0.311495
12                 0.898291  0.017009  0.771947
13                 0.310711  0.997464  0.398520
14                 0.507288  0.058361  0.864464
15                 0.215043  0.144644  0.436364
16                 0.183891  0.544676  0.280621
17                 0.391560  0.594126  0.436677

In [51]:
new_coffee_levels.reset_index()

hour coffee_amount                    
coffee_carafe                  x         y         z
0                8      0.816164  0.189297  0.999264
1                9      0.451018  0.521502  0.915990
2               10      0.843279  0.023163  0.144928
3               11      0.335533  0.235529  0.311495
4               12      0.898291  0.017009  0.771947
5               13      0.310711  0.997464  0.398520
6               14      0.507288  0.058361  0.864464
7               15      0.215043  0.144644  0.436364
8               16      0.183891  0.544676  0.280621
9               17      0.391560  0.594126  0.436677

Is this the best shape for the data?
-seems to be

Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [52]:
cake_recipes = pd.read_csv('untidy-data/cake_recipes.csv')

In [53]:
cake_recipes.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [54]:
cake_recipes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   recipe:position  8 non-null      object 
 1   225              8 non-null      float64
 2   250              8 non-null      float64
 3   275              8 non-null      float64
 4   300              8 non-null      float64
dtypes: float64(4), object(1)
memory usage: 448.0+ bytes


Tidy the data as necessary.

In [55]:
cake_recipes.shape

(8, 5)

In [56]:
cake_recipes = cake_recipes.rename(columns={"recipe:position": "recipe"})

In [57]:
cake_recipes

recipe        225        250        275        300
0  a:bottom  61.738655  53.912627  74.414730  98.786784
1     a:top  51.709751  52.009735  68.576858  50.228470
2  b:bottom  57.095320  61.904369  61.196980  99.248541
3     b:top  82.455004  95.224151  98.594881  58.169349
4  c:bottom  96.470207  52.001358  92.893227  65.473084
5     c:top  71.306308  82.795477  92.098049  53.960273
6  d:bottom  52.799753  58.670419  51.747686  56.183110
7     d:top  96.873178  76.101363  59.571620  50.971626

In [58]:
new_cake_recipes = cake_recipes.recipe.str.split(':', expand=True)
new_cake_recipes.columns = ['Recipes', 'Position']
new_cake_recipes = pd.concat([cake_recipes, new_cake_recipes], axis=1).drop(columns='recipe')


In [59]:
new_cake_recipes

225        250        275        300 Recipes Position
0  61.738655  53.912627  74.414730  98.786784       a   bottom
1  51.709751  52.009735  68.576858  50.228470       a      top
2  57.095320  61.904369  61.196980  99.248541       b   bottom
3  82.455004  95.224151  98.594881  58.169349       b      top
4  96.470207  52.001358  92.893227  65.473084       c   bottom
5  71.306308  82.795477  92.098049  53.960273       c      top
6  52.799753  58.670419  51.747686  56.183110       d   bottom
7  96.873178  76.101363  59.571620  50.971626       d      top

In [30]:
# df.melt(id_vars=['a', 'b'], value_vars=['x', 'y'], var_name='original_column_name', value_name='original_value')
# new_cake_recipes.melt(id_vars=['Recipes'])

In [65]:
cake_recipes = new_cake_recipes.melt(id_vars=['Recipes', 'Position'])

In [66]:
cake_recipes

Recipes Position variable      value
0        a   bottom      225  61.738655
1        a      top      225  51.709751
2        b   bottom      225  57.095320
3        b      top      225  82.455004
4        c   bottom      225  96.470207
5        c      top      225  71.306308
6        d   bottom      225  52.799753
7        d      top      225  96.873178
8        a   bottom      250  53.912627
9        a      top      250  52.009735
10       b   bottom      250  61.904369
11       b      top      250  95.224151
12       c   bottom      250  52.001358
13       c      top      250  82.795477
14       d   bottom      250  58.670419
15       d      top      250  76.101363
16       a   bottom      275  74.414730
17       a      top      275  68.576858
18       b   bottom      275  61.196980
19       b      top      275  98.594881
20       c   bottom      275  92.893227
21       c      top      275  92.098049
22       d   bottom      275  51.747686
23       d      top      275  59.571620
24       a   bottom      300  98.786784
25       a      top      300  50.228470
26       b   bottom      300  99.248541
27       b      top      300  58.169349
28       c   bottom      300  65.473084
29       c      top      300  53.960273
30       d   bottom      300  56.183110
31       d      top      300  50.971626

In [67]:
cake_recipes = cake_recipes.rename(columns={"value": "taste_score"})

In [69]:
cake_recipes = cake_recipes.rename(columns={"variable": "temp"})

In [70]:
cake_recipes.head()

Recipes Position temp  taste_score
0       a   bottom  225    61.738655
1       a      top  225    51.709751
2       b   bottom  225    57.095320
3       b      top  225    82.455004
4       c   bottom  225    96.470207

Which recipe, on average, is the best? recipe b



In [71]:
cake_recipes.groupby('Recipes').taste_score.mean()

Recipes
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: taste_score, dtype: float64

Which oven temperature, on average, produces the best results? 275



In [72]:
cake_recipes.groupby('temp').taste_score.mean()

temp
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: taste_score, dtype: float64

Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degree



In [75]:
cake_recipes.sort_values(by='taste_score').tail()

Recipes Position temp  taste_score
4        c   bottom  225    96.470207
7        d      top  225    96.873178
19       b      top  275    98.594881
24       a   bottom  300    98.786784
26       b   bottom  300    99.248541